src/Processors/Transforms/AggregatingTransform.cpp

```diff
@@ -280,16 +352,67 @@ class ConvertingAggregatedToChunksTransform final : public IProcessor
         return Status::PortFull;
     }
 
+    /// proton : starts
+    /// Read all sources and try to push current processed bucket.
+    IProcessor::Status prepareTwoLevelShuffled()
+    {
+        auto & output = outputs.front();
+
+        size_t new_processed = 0;
+        for (auto & input : inputs)
+        {
+            if (!input.isFinished() && input.hasData())
+            {
+                shuffled_chunks.push_back(input.pull());
```

Comment:
NOTE, for streaming aggregating, we need to use `pull(/*set_not_needed*/true)`, since the inputs can receive heartbeat chunk, then trigger upstream again, next loop above actions, which will cause `pipeline stuck`, there are some reasons:
1) The upstream is triggered first, resulting in no thread calling the downstream processor.
2) Missing propagate heartbeat chunk can also cause pipeline stuck.
But it's work in historical aggregating.

Reply:
This is pure historical data. Streaming/AggregatingTransform.cpp is not covered yet